TickSpec/FeatureParser.fs

```diff
@@ -1,92 +1,131 @@
+            ) (Some (List.Empty, Map.empty))
+
+        examples
+        |> Seq.map (fun exampleBlock ->
+            exampleBlock.Table.Header |> List.sort,
+            exampleBlock.Table.Rows
+            |> Seq.map (fun row ->
+                Seq.zip exampleBlock.Table.Header row
+                |> Seq.fold (fun map (header, value) ->
+                        match Map.tryFind header map with
+                        | Some x ->
+                            let m = sprintf "A single header was specified multiple times in an example block starting at row %d" exampleBlock.LineNumber
+                            ParseException(m, Some exampleBlock.LineNumber) |> raise
+                        | None -> map |> Map.add header value
+                    ) Map.empty,
+                exampleBlock.Tags
             )
         )
-    let bullets =
-        line.Bullets
-        |> Option.map (fun bullets -> bullets |> Array.map replace)
-    (scenario,n,tags,{line with Table=table;Bullets=bullets},step)
-
-/// Appends shared examples to scenarios as examples
-let internal appendSharedExamples (sharedExamples:Table[]) scenarios  =
-    if Seq.length sharedExamples = 0 then
-        scenarios
-    else
-        scenarios |> Seq.map (function
-            | scenarioName,tags,steps,None ->
-                scenarioName,tags,steps,Some(sharedExamples)
-            | scenarioName,tags,steps,Some(exampleTables) ->
-                scenarioName,tags,steps,Some(Array.append exampleTables sharedExamples)
+        // Union tables with the same columns
+        |> Seq.groupBy (fun (h,_) -> h)
+        |> Seq.map (fun (header,tables) ->
+            header, tables |> Seq.collect (fun (_, rows) -> rows) |> Seq.toList)
+        |> Seq.toList
+        // Cross-join tables with different columns
+        |> combinations
+        |> Seq.map processRow
+        |> Seq.choose id
+        |> Seq.groupBy (fun (_,r) -> r)
+        |> Seq.map (fun (row, taggedRows) ->
+            taggedRows
+            |> Seq.fold (fun tags (t,r) ->
+                Seq.append tags t
+            ) Seq.empty
+            |> Seq.distinct
+            |> Seq.toList,
+            row |> Map.toList
         )
 
-/// Parses lines of feature
-let parseFeature (lines:string[]) =
-    let toStep (_,_,_,line,step) = step,line
-    let featureName,background,scenarios,sharedExamples = parseBlocks lines
+    let createStep combination step =
+        let processedStep =
+            match step.Step with
+            | GivenStep s -> replace combination s |> GivenStep
+            | WhenStep s -> replace combination s |> WhenStep
+            | ThenStep s  -> replace combination s |> ThenStep
+
+        let bullets, table, doc =
+            match step.Item with
+            | Some (BulletsItem b) -> Some (b |> List.toArray), None, None
+            | Some (TableItem t) -> None, Some (new Table(t.Header |> List.toArray, t.Rows |> Seq.map List.toArray |> Seq.toArray)), None
```

Comment:
break this out  - in general if you find this many parens, its a signal to split into temps

Reply:
Changed, thanks.